Implement basic JavaPoet interop module

ZacSweers opened the pull request:
Resolves #417 and is based on the implementation I wrote in the past that @shaishavgandhi shared in the issue via https://gist.github.com/ShaishavGandhi/097033cc528ae25741186973e4d36ce4

This'll be pretty helpful for processors like Room that support generating both Java and Kotlin. I know we weren't sure a few years ago when the issue was first created, but I think KSP's dual support will create new avenues for this where this becomes helpful. We also have a bit of a stronger interop artifacts story now too.

TODO:
- [x] Tests (see below)
- [x] `WildcardTypeName` interop
- [x] Do we want to use an experimenta/preview annotation for these?
- [x] What about other Kotlin intrinsics? Should we cover the set contained in KotlinPoet itself (beyond just `LIST`, `SET`, etc)?

ZacSweers commented:
@ShaishavGandhi could you share the tests that were associated with these interop extensions?

Egorand commented on interop/javapoet/src/main/kotlin/com/squareup/kotlinpoet/javapoet/typeAliases.kt:
```diff
@@ -0,0 +1,36 @@
+ */
+package com.squareup.kotlinpoet.javapoet
+
+/*
+ * Useful typealiases for colliding names
+ */
```
nit (I believe as is this would be attached as doc for `KTypeName` which seems wrong):
```suggestion
// Useful typealiases for colliding names
```

ZacSweers replied:
It actually won't be attached to any elements if it's just a single star statement like this. It only becomes a full kdoc if you have two stars (`/**`)

Egorand replied:
Ah, true, nvm then!

Egorand commented on interop/javapoet/src/main/kotlin/com/squareup/kotlinpoet/javapoet/k2jInterop.kt:
```diff
@@ -0,0 +1,98 @@
+import com.squareup.kotlinpoet.SHORT
+import com.squareup.kotlinpoet.STRING
+
+public fun KClassName.toJClassName(shouldBox: Boolean = false): JTypeName {
```
nit: is `boxIfPrimitive` a better name than `shouldBox`?

ZacSweers replied:
5d15587

Egorand commented on interop/javapoet/gradle.properties:
```diff
@@ -0,0 +1,4 @@
+POM_ARTIFACT_ID=kotlinpoet-ksp
+POM_NAME=KotlinPoet (KSP Interop)
```
Copy pasta!

ZacSweers replied:
🙈

ZacSweers replied:
cf35d5a

Egorand reviewed (approved):
Sweet! Yep, tests would be great.

ZacSweers commented:
@Egorand thoughts on the last two TODO items in the desciption?

Egorand commented:
>  Do we want to use an experimenta/preview annotation for these?

I'd say yes, will give us more flexibility to change the API if needed.

>  What about other Kotlin intrinsics? Should we cover the set contained in KotlinPoet itself (beyond just LIST, SET, etc)?

Can definitely do this in followup PRs, and postpone until there are use cases for it. Another idea would be to open an issue and accept external contributions for these features.

ShaishavGandhi reviewed (approved):
LGTM! Updated the gist with the tests.

ZacSweers commented:
Added a bunch of tests (and fixed some bugs they revealed!), the preview annotation, and more intrinsics coverage. I think this is a good enough start!
